snuba/query/processors/project_rate_limiter.py

```diff
@@ -0,0 +1,51 @@
+from snuba.clickhouse.query_dsl.accessors import get_project_ids_in_query_ast
+from snuba.query.logical import Query
+from snuba.query.processors import QueryProcessor
+from snuba.request.request_settings import RequestSettings
+from snuba.state import get_configs
+from snuba.state.rate_limit import PROJECT_RATE_LIMIT_NAME, RateLimitParameters
+
+
+class ProjectRateLimiterProcessor(QueryProcessor):
+    """
+    If there isn't already a rate limiter on a project, search the top level
+    conditions for project IDs using the given project column name and add a
+    rate limiter for them.
+    """
+
+    def __init__(self, project_column: str) -> None:
+        self.project_column = project_column
+
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        # If the settings don't already have a project rate limit, add one
+        existing = request_settings.get_rate_limit_params()
+        for ex in existing:
+            if ex.rate_limit_name == PROJECT_RATE_LIMIT_NAME:
+                return
+
+        project_ids = get_project_ids_in_query_ast(query, self.project_column)
+        if not project_ids:
+            return
+
+        project_id = project_ids.pop()
```

Comment:
Please add a todo that this needs to change according to the backlog task we have. (no need to link the task).